corehq/messaging/scheduling/models/timed_schedule.py

```diff
@@ -0,0 +1,201 @@
+    MONTHLY = -1
+
+    schedule_length = models.IntegerField()
+    total_iterations = models.IntegerField()
+
+    @property
+    @memoized
+    def memoized_events(self):
+        """
+        This is named with a memoized_ prefix to be clear that it should only be used
+        when the event set is not changing.
+        """
+        return list(self.timedevent_set.order_by('order'))
+
+    def set_first_event_due_timestamp(self, instance, start_date=None):
+        """
+        If start_date is None, we set it automatically ensuring that
+        self.next_event_due does not get set in the past for the first
+        event.
+        """
+        if start_date:
+            instance.start_date = start_date
+        else:
+            instance.start_date = ServerTime(util.utcnow()).user_time(instance.timezone).done().date()
+
+        self.set_next_event_due_timestamp(instance)
+
+        if (
+            not self.schedule_length == self.MONTHLY and
+            not start_date and
+            instance.next_event_due < util.utcnow()
+        ):
+            instance.start_date += timedelta(days=1)
+            instance.next_event_due += timedelta(days=1)
+
+    def get_local_next_event_due_timestamp(self, instance):
+        current_event = self.memoized_events[instance.current_event_num]
+
+        days_since_start_date = (
+            ((instance.schedule_iteration_num - 1) * self.schedule_length) + current_event.day
+        )
+
+        return datetime.combine(
+            instance.start_date + timedelta(days=days_since_start_date),
+            current_event.time
+        )
+
+    def get_local_next_event_due_timestamp_for_monthly_schedule(self, instance):
+        target_date = None
+
+        while target_date is None:
+            current_event = self.memoized_events[instance.current_event_num]
+
+            if current_event.day < -28 or current_event.day == 0 or current_event.day > 31:
+                # Negative days are days from the end of the month, with -1
+                # being the last day of the month. We don't allow this going
+                # past -28 since it's not very useful to do so, and imposing
+                # this restriction lets us make the assumption that we can
+                # always schedule a negative day.
+                raise InvalidMonthlyScheduleConfiguration("Day must be between -28 and 31, and not be 0")
+
+            year_offset = (instance.schedule_iteration_num - 1) / 12
+            month_offset = (instance.schedule_iteration_num - 1) % 12
+
+            year = instance.start_date.year + year_offset
+            month = instance.start_date.month + month_offset
+
+            days_in_month = calendar.monthrange(year, month)[1]
+            if current_event.day > 0:
+                if current_event.day > days_in_month:
+                    # If the day refers to a date that is not possible to schedule
+                    # (for example, February 30th), just move to the next month.
```

Comment:
so actually the behavior here is that if someone starts a monthly schedule on the 30th, it would skip the event for February.

if someone wanted to schedule on the last day of the month, they could enter a -1 for the day. we don't have a way to do "last day of the month" in our ui spec since it's not a requirement yet, but i imagine some day that would be a nice thing to add and the backend is ready for it now.